### ISSUE TITLE
Proposal: do not delete packages from branches

### ISSUE THREAD

**Issue by mtojek** [elastic/package-storage#212]
Hi,

I have on my roadmap a plan for preparing a release management solution and I was wondering if we can simplify some actions here.

I propose to shape a Git flow here and do not delete already promoted packages from branches. The overall look of the package-storage will change a bit:

Currently -
snapshot: A.34
staging: A.33
production: A.30

Would be -
snapshot: A.30, A.33, A.34
staging: A.30, A.33
production: A.30

I bet it will simplify releasing new changes. I'd like to end up with a single PR that promotes changes from snapshot branch to the staging branch. One click and we're done. We could automatically kickoff the package-registry deployment on commit push.

**Comment by mtojek** [comment]
FYI @ruflin @ycombinator @jonathan-buttner

**Comment by ycombinator** [comment]
I imagine with this proposed change the three registry Docker images — `snapshot`, `staging`, and `production` end up having a 1-1 relationship with the corresponding branches in the `package-storage` repo? If so, I like it — it certainly simplifies the understanding of the registry stages in relation to the `package-storage` branches, it simplifies the promotion process like you noted, and it would also simplify the checks done by the `mage updatePackageStorage` target in the `integrations` repo.

So 👍 from my end.

**Comment by jonathan-buttner** [comment]
This sounds good to me. I'd still like to support for development straight to the `staging` branch. For endpoint development I've mostly been skipping the `snapshot` branch when releasing the package and being going straight to `staging`. So if the release script to work with a package starting from `snapshot -> staging` and `staging -> production` that'd be great!

**Comment by ruflin** [comment]
My initial intention with having only the current package for example in staging is to make it obvious which packages are currently there for testing and not release yet. If all of them are there, this becomes tricky.

My thinking around snapshot containing all the other packages is to make sure the user gets the full experience there. Some packages might otherwise be missing, for example the system package. This becomes especially a problem if we push directly to staging or prod (which I think we should allow as suggested by @jonathan-buttner )

What I fully agree, pushing packages is too much effort at the moment. It should be a simple command or similar and not involve much user interaction.

I'm keep getting pack to the "direct" commit approach for staging and snapshot instead of PRs. This is until we have an actual API we can push against.

**Comment by mtojek** [comment]
> I'm keep getting pack to the "direct" commit approach for staging and snapshot instead of PRs. This is until we have an actual API we can push against.

I fundamentally disagree with the approach of pushing changes that haven't been verified by the CI. This is not a box product, we must tend to manage a full CI/CD here. From ops perspective it impacts visibility much. Yesterday we deleted a couple of packages - imagine tracking/referring to this change without the PR. PRs also attract to discussions and researches around potential issues.

> This becomes especially a problem if we push directly to staging or prod (which I think we should allow as suggested by @jonathan-buttner )

That's why shouldn't do that! Every change should follow the pipeline no matter what. If you tell me to jump over snapshot and staging every second time, I would say that we don't need these stages. Direct deployment to prod should have a really good justification and should be performed only in emergency cases.

**Comment by ruflin** [comment]
> I fundamentally disagree with the approach of pushing changes that haven't been verified by the CI. 
@mtojek I'm glad you do, me too. I think we are on the same page about the CI / CD pipeline.

> Direct deployments
I think you are right here. The main reason we do it at the moment is because we have too many emergencies and the pipeline is not optimal yet.

@mtojek TBH I think we are mostly on the same page. My guess that the above potential just becomes a "given" as soon as the proper CI / CD is in place.

**Comment by ruflin** [comment]
@mtojek I've been thinking about this again. At the moment the process is not ideal and this will simplify it. I would say lets try it out?

**Comment by mtojek** [comment]
Once I'm done with changes in https://github.com/elastic/integrations/pull/224 I'll will switch to this topic. I was also waiting for deploying changes around data_stream and as we did it yesterday, I don't find any other blockers.

**Comment by mtojek** [comment]
Another idea: let's keep the content in branches snapshot, staging, production equal, I mean exactly the same. Then we'll end up with a PRs like: merge snapshot into staging, merge staging into prod.

Docker tags can be determined by branch names (distribution:<branch_name>).

At the moment, we need 3 custom PRs to different branches to update EPR. If we agree on the idea, it would require also 3 PRs, but only the first one is crucial, the other ones are only syncs.

**Comment by ruflin** [comment]
* What about `preprelease` packages?  I assume they should never end up in `staging` or `prod`
* What if we put a package into snapshot and decide to pull it again?
* How would promoting one package from staging to prod work without promoting all of them?

**Comment by mtojek** [comment]
> What about preprelease packages? I assume they should never end up in staging or prod

Which packages are prerelease? Do you refer to snapshot and staging packages? I believe in this setup we can delete them.

> What if we put a package into snapshot and decide to pull it again?

What do you mean by pull it again?

> How would promoting one package from staging to prod work without promoting all of them?

I meant untouched merge from snapshot to prod when we want to move all packages at once. In cases where we need to move some packages only, we can edit the PR or simply rsync.

**Comment by ruflin** [comment]
* Prerelease: None yet, but might be coming: https://github.com/elastic/package-storage/issues/252
* Pull: Should have been more specific, delete a package because it was just to try it out

The last point is interesting. What is our assumption on how packages are deployed in the future? Today we often do bulk deployments because we do bulk changes. But I do not expect that this is how things work in the future. I expect a dev to work on a package, put it into snapshot/staging, test it, push it to prod. This dev does not care about the other packages. I expect this to become the norm.

**Comment by ruflin** [comment]
Additional thought: What if we split up the environments in 2 parts:

* staging / production: All packages go through this cycle
* snapshot: This is additional and must not be always used. It might also work a bit different by pulling in packages from other repos etc.

**Comment by mtojek** [comment]
> Pull: Should have been more specific, delete a package because it was just to try it out

It should never happen, but it's a matter solution flexibility. I don't expect the whole system to collapse if we do such thing.

> snapshot: This is additional and must not be always used. It might also work a bit different by pulling in packages from other repos etc.

There would be a risk that this will become a playground which "always" doesn't work or somebody deployed some crappy revisions there.

> What is our assumption on how packages are deployed in the future?
> But I do not expect that this is how things work in the future

My goal is to cure us of current issues and don't focus on long-term plans. I expect the package-storage design to be changed/replaced in approx. 3 month.

**Comment by jonathan-buttner** [comment]
> Another idea: let's keep the content in branches snapshot, staging, production equal, I mean exactly the same. Then we'll end up with a PRs like: merge snapshot into staging, merge staging into prod.

Hmm I think we should treat `snapshot` as a special place that will have packages that will never make it to `staging` or `production`. For example this happened recently for the endpoint team:

One of our developers needed to push an interim/prerelease package to an environment so that a new docker image would be built that we could pin to in our kibana integration tests. This interim package will likely never be released (another package that supersedes it will be released, or we may remove the feature in the package all together). The `snapshot` branch is a great place to push these prerelease packages because developers working on kibana master automatically connect to the `snapshot` registry by default. So when we add new mappings or a feature in the endpoint package that isn't quite ready for production but is needed so the rest of our development team can use it, we can throw it temporarily in `snapshot` to share it across the team.

I think we should have some script that **can** promote a package from `snapshot` through the rest of the process but I'd also be ok with having `staging` be the starting point for a well tested package developed locally that never needed to go to `snapshot`. If we feel like **all** packages need to go through `snapshot` first, that's fine, but I think we should support the case where a package will never leave `snapshot`.


> * What about `preprelease` packages?  I assume they should never end up in `staging` or `prod`
Yeah my thought is that prerelease packages will never get to `staging` or `prod`
> * What if we put a package into snapshot and decide to pull it again?
Yeah I agree here as well, we should be able to remove a package from `snapshot`, or just leave it in `snapshot` forever.

> * How would promoting one package from staging to prod work without promoting all of them?
Yeah I think we should support promoting one to many packages from one environment to the next.

**Comment by mtojek** [comment]
Aaa, I see where the problem lies. We used to keep development inside the integrations repo, so that all prerelease packages don't leave the repo. We release only these ones that should be pushed.

I'm not quite convinced about the way you used to publish to snapshot and potentially delete it. Maybe it's because I consider the snapshot as same important stage of the release process as others, so I believe we shouldn't delete already added packages. I wonder if we need an option to promote the latest version only.

I think this is a good moment to redefine goals of particular stages - which packages should belong to snapshot, staging and production. Which ones can be removed, promoted, etc.

**Comment by mtojek** [comment]
Regarding the script, this is the task I'd like to focus on (incl. implementation not just design), so I need to introduce potential modifications before I will start building tools.

**Comment by jonathan-buttner** [comment]
> Aaa, I see where the problem lies. We used to keep development inside the integrations repo, so that all prerelease packages don't leave the repo. We release only these ones that should be pushed.

Yeah unfortunately I think prerelease packages will need to exist in `package-storage`. This is because we need a docker image built and published that we can use in our Kibana integration. This will likely happen before we plan on releasing a package to production.

> 
> I'm not quite convinced about the way you used to publish to snapshot and potentially delete it. Maybe it's because I consider the snapshot as same important stage of the release process as others, so I believe we shouldn't delete already added packages.

I'm fine with not deleting a prerelease package. The way I'm thinking of using prereleases is incrementing a number when we publish them so having multiple prereleases in `snapshot` shouldn't be an issue.

> I wonder if we need an option to promote the latest version only.

I think that'd be helpful. If we have `0.14.0-dev.5` and `0.14.0`, the promotion script should only promote `0.14.0`. I believe that is how semver works though, so it shouldn't be an issue.

> 
> I think this is a good moment to redefine goals of particular stages - which packages should belong to snapshot, staging and production. Which ones can be removed, promoted, etc.

I also think we should define what the purpose of each environment is. I see the `snapshot` environment being a place where developers can put a package that other developers can access while running Kibana on `master`. It is a place to do incremental testing and place to put a prerelease package that is not intended to ever be released to production. I think once a package is ready to be released it _could_ go straight to `staging` for a last bit of testing.

If a package that is intended to be released to production goes first to `snapshot` and testing is completed there, what further testing is going to happen in `staging`? Would it be additional CI testing across multiple kibana versions or something?

I guess my question is if **all** packages have to go to `snapshot` first and then `staging` before `production` what things are we checking/testing differently between when the package is in `snapshot` vs when it is `staging`?

**Comment by mtojek** [comment]
> If a package that is intended to be released to production goes first to snapshot and testing is completed there, what further testing is going to happen in staging? Would it be additional CI testing across multiple kibana versions or something?

We used to handle this a bit differently. We parked all integrations in staging, tested them there and left until a green light for the release to prod. This can be anything starting from a notice, blog post to official announcements.

> I guess my question is if all packages have to go to snapshot first and then staging before production what things are we checking/testing differently between when the package is in snapshot vs when it is staging?

I had a similar confusion :) What do you think about merging snapshot into staging and keep two stages.

**Comment by ruflin** [comment]
I personally still see value in the three branches. I think we are mostly aligned on what staging and production are for:

* staging: Packages ready to be tested and very likely to be released as is. This is a "release candidate" and where testing happens.
* production: Released packages

Where I think `snapshot` is really valuable is to really have snapshot builds of packages. I compare it to what we have as `master` in our repositories. The build is always green but features might be incomplete but still already working. It allows others to already start trying out a feature, in our case a package. I think a good example is the synthetics package here: https://github.com/elastic/integrations/pull/241 Elastic Agent master supports it but none of the other branches. Also the way the package is at the moment, it will never be shipped to production but we need the current state to start iterating on it.

Applying the above to the `integrations` repo. Currently most PR's with changes to a package also bump the version. Also CI is already running the packages and the packages are ready for release. Because of this, most integrations packages from my perspective could go directly to `staging`.

For the development cycle of `endpoint` I'm wondering where `0.14.0-dev.5` should go, staging or snapshot. Does each commit to the endpoint package created a new dev build? Who is expected to test these dev builds? If we expect a testing team to test each of these builds, I would expect them to go to staging directly. But if it is more that someone working with Kibana and Endpoint UI can directly use this package on Kibana `master` after the change was made, I rather put it into snapshot and only when the feature is completed and backported to 7.x, the package is pushed to staging and made ready for testing.

++ on having a script to promoted the latest package only.

**Comment by mtojek** [comment]
> ++ on having a script to promoted the latest package only.

I think in the initial version I will adjust to the current design, not to introduce too many changes at once. We can iterate on this later on.

Regarding questions regarding the package development cycle based on the "endpoint" package, I think these are valid concerns that need to be addressed in package-storage 2.0.

**Comment by jonathan-buttner** [comment]
> I personally still see value in the three branches. I think we are mostly aligned on what staging and production are for:
> 
> * staging: Packages ready to be tested and very likely to be released as is. This is a "release candidate" and where testing happens.
> * production: Released packages

I agree, I think that's a good summary of the purpose of these environments.

> Where I think `snapshot` is really valuable is to really have snapshot builds of packages. I compare it to what we have as `master` in our repositories. The build is always green but features might be incomplete but still already working. It allows others to already start trying out a feature, in our case a package. I think a good example is the synthetics package here: [elastic/integrations#241](https://github.com/elastic/integrations/pull/241) Elastic Agent master supports it but none of the other branches. Also the way the package is at the moment, it will never be shipped to production but we need the current state to start iterating on it.

I agree here too. This is also how the endpoint team will use the `snapshot` environment.

> For the development cycle of `endpoint` I'm wondering where `0.14.0-dev.5` should go, staging or snapshot. Does each commit to the endpoint package created a new dev build? Who is expected to test these dev builds? If we expect a testing team to test each of these builds, I would expect them to go to staging directly. But if it is more that someone working with Kibana and Endpoint UI can directly use this package on Kibana `master` after the change was made, I rather put it into snapshot and only when the feature is completed and backported to 7.x, the package is pushed to staging and made ready for testing.
> 

I think `0.14.0-dev.5` should go to `snapshot`. No each commit does not create a new dev build. A developer has to intentionally publish a dev build for remote testing. The developer working on the feature is expected to test the dev builds. Yeah the purpose is so someone working with Kibana and Endpoint UI can directly use the package on Kibana `master`.

> only when the feature is completed and backported to 7.x, the package is pushed to staging and made ready for testing.

Yep I agree here too.

> ++ on having a script to promoted the latest package only.

I'm fine with this. I think as mentioned before the script should allow promoting a single package at a time (and maybe an option to promote multiple packages?).

**Comment by mtojek** [comment]
I suppose I can close the issue. It's a bit stale now.
